src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs

```diff
@@ -0,0 +1,192 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DataAI.NuGetRecommender.Contracts;
+using Microsoft.VisualStudio.Shell;
+using NuGet.Common;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Represents a package feed which recommends packages based on currently loaded project info
+    /// </summary>
+    public class RecommenderPackageFeed : IPackageFeed
+    {
+        public bool IsMultiSource => false;
+
+        private readonly SourceRepository _sourceRepository;
+        private readonly IEnumerable<PackageCollectionItem> _installedPackages;
+        private readonly Dictionary<string, VersionRange> _dependentPackages;
+        private readonly IEnumerable<string> _targetFrameworks;
+        private readonly IPackageMetadataProvider _metadataProvider;
+        private readonly Common.ILogger _logger;
+
+        public (string modelVersion, string vsixVersion) VersionInfo = (modelVersion: (string)null, vsixVersion: (string)null);
+
+        IVsNuGetPackageRecommender NuGetRecommender { get; set; }
+
+        public RecommenderPackageFeed(
+            SourceRepository sourceRepository,
+            IEnumerable<PackageCollectionItem> installedPackages,
+            Dictionary<string, VersionRange> dependentPackages,
+            IEnumerable<string> targetFrameworks,
+            IPackageMetadataProvider metadataProvider,
+            Common.ILogger logger)
+        {
+            if (sourceRepository == null)
+            {
+                throw new ArgumentNullException(nameof(sourceRepository));
+            }
+            _sourceRepository = sourceRepository;
+
+            if (installedPackages == null)
+            {
+                throw new ArgumentNullException(nameof(installedPackages));
+            }
+            _installedPackages = installedPackages;
+
+            if (dependentPackages == null)
+            {
+                throw new ArgumentNullException(nameof(dependentPackages));
+            }
+            _dependentPackages = dependentPackages;
+
+            if (targetFrameworks == null)
+            {
+                throw new ArgumentNullException(nameof(targetFrameworks));
+            }
+            _targetFrameworks = targetFrameworks;
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+            _metadataProvider = metadataProvider;
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            _logger = logger;
+
+            // Get NuGet package recommender service, or null if it is not available
+            NuGetRecommender = Package.GetGlobalService(typeof(SVsNuGetRecommenderService)) as IVsNuGetPackageRecommender;
```

Comment:
This needs the UI thread. 
We should not use thread affinitized methods in constructors. 

This should be either loaded through MEF/lazily.

Reply:
Working on this. Will let you know when I have a fix.

Reply:
I updated this to use AsyncLazy